From e80381ac828c13a6bc9aaa7bc2f6de7d1925c655 Mon Sep 17 00:00:00 2001
From: Anson Huang <Anson.Huang@nxp.com>
Date: Thu, 8 Dec 2016 22:40:35 +0800
Subject: [PATCH 1329/5242] MLK-13580 ARM: imx: update DGO register to SNVS
 domain on i.MX7ULP

commit  ee0c28df5e4a1dfd2505a5a17a95e8ccec0cfe96 from
https://source.codeaurora.org/external/imx/linux-imx.git

On i.MX7ULP, the resume entry and parameter are saved
in DGO_GP registers, it has two power domains, one
is normal domain, the other is SNVS domain, if M4 also
enters VLLS mode, DGO needs to be updated into SNVS
domain to avoid power lost and lead to resume fail.

Signed-off-by: Anson Huang <Anson.Huang@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 arch/arm/mach-imx/suspend-imx7ulp.S |   13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/arch/arm/mach-imx/suspend-imx7ulp.S b/arch/arm/mach-imx/suspend-imx7ulp.S
index 40a3ee3..bb250c6 100644
--- a/arch/arm/mach-imx/suspend-imx7ulp.S
+++ b/arch/arm/mach-imx/suspend-imx7ulp.S
@@ -283,6 +283,19 @@ ENTRY(imx7ulp_suspend)
 	/* store physical resume addr and pm_info address. */
 	str	r9, [r11, #DGO_GPR3]
 	str	r1, [r11, #DGO_GPR4]
+	ldr	r7, [r11, #DGO_CTRL0]
+	orr	r7, r7, #0xc
+	str	r7, [r11, #DGO_CTRL0]
+wait_dgo:
+	ldr	r7, [r11, #DGO_CTRL0]
+	and	r7, r7, #0x18000
+	cmp	r7, #0x18000
+	bne	wait_dgo
+
+	ldr	r7, [r11, #DGO_CTRL0]
+	orr	r7, r7, #0x18000
+	bic	r7, r7, #0xc
+	str	r7, [r11, #DGO_CTRL0]
 
 	disable_l1_dcache
 
-- 
1.7.9.5

